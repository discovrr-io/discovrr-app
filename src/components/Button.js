import React from 'react';
// import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import { Text, TouchableHighlight } from 'react-native';

import { colors, typography, values } from '../constants';

const Button = ({
  title,
  primary = false,
  size = 'big',
  disabled = false,
  onPress = () => {},
  ...props
}) => {
  const stateStyle = primary ? primaryStyle : secondaryStyle;
  const currentStyle = disabled ? stateStyle.disabled : stateStyle.default;

  const isBig = size === 'big';
  const sizeStyle = isBig ? bigStyle : smallStyle;

  return (
    <TouchableHighlight
      disabled={disabled}
      underlayColor={primary ? colors.accentFocused : colors.lightGray}
      onPress={onPress}
      style={[currentStyle, sizeStyle, props.style]}>
      <Text style={isBig ? stateStyle.text : stateStyle.textSmall}>
        {title}
      </Text>
    </TouchableHighlight>
  );
};

// Button.propTypes = {
//   title: PropTypes.string.isRequired,
//   primary: PropTypes.bool,
//   disabled: PropTypes.bool,
//   onPress: PropTypes.func,
// };

const buttonStyles = {
  alignItems: 'center',
  borderWidth: values.border.thin,
  justifyContent: 'center',
};

const smallStyle = {
  height: 28,
  borderRadius: values.radius.md,
  padding: values.spacing.sm,
};

const bigStyle = {
  height: 50,
  borderRadius: values.radius.lg,

  padding: values.spacing.md,
};

const textStyle = {
  fontWeight: '700',
};

const bigTextStyle = {
  ...textStyle,
  fontSize: typography.size.lg,
};

const smallTextStyle = {
  ...textStyle,
  fontSize: typography.size.sm,
};

const primaryStyle = StyleSheet.create({
  text: {
    ...bigTextStyle,
    color: colors.white,
  },
  textSmall: {
    ...smallTextStyle,
    color: colors.white,
  },
  default: {
    ...buttonStyles,
    backgroundColor: colors.accent,
    borderColor: colors.accent,
  },
  disabled: {
    ...buttonStyles,
    backgroundColor: colors.accentDisabled,
    borderColor: colors.accentDisabled,
  },
});

const secondaryStyle = StyleSheet.create({
  text: {
    ...bigTextStyle,
    color: colors.gray,
  },
  textSmall: {
    ...smallTextStyle,
    color: colors.gray,
  },
  default: {
    ...buttonStyles,
    backgroundColor: colors.white,
    borderColor: colors.gray,
  },
  disabled: {
    ...buttonStyles,
    backgroundColor: colors.white,
    borderColor: colors.lightGray,
  },
});

export default Button;
