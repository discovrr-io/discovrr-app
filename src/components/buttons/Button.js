import React from 'react';
import PropTypes from 'prop-types';
import { Text, TouchableHighlight } from 'react-native';

import { colors } from '../constants';
import styles from './styles';

const Button = ({
  title,
  primary = false,
  size = 'big',
  disabled = false,
  transparent = false,
  onPress = () => {},
  ...props
}) => {
  const stateStyle = primary
    ? styles.primaryStyle
    : transparent
    ? styles.transparentStyle
    : styles.secondaryStyle;
  const currentStyle = disabled ? stateStyle.disabled : stateStyle.default;

  const isBig = size === 'big';
  const sizeStyle = isBig ? styles.bigStyle : styles.smallStyle;

  return (
    <TouchableHighlight
      disabled={disabled}
      underlayColor={primary ? colors.accentFocused : colors.lightGray}
      onPress={onPress}
      style={[currentStyle, sizeStyle, props.style]}>
      <Text style={isBig ? stateStyle.text : stateStyle.textSmall}>
        {title}
      </Text>
    </TouchableHighlight>
  );
};

Button.propTypes = {
  title: PropTypes.string.isRequired,
  primary: PropTypes.bool,
  size: PropTypes.oneOf(['big', 'small']),
  disabled: PropTypes.bool,
  transparent: PropTypes.bool,
  onPress: PropTypes.func,
};

export default Button;
