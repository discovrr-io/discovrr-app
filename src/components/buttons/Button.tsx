import React, { useMemo } from 'react';

import { ButtonBase } from './ButtonBase';
import {
  CommonButtonProps,
  makeButtonStyles,
  makeContainedButtonColorStyles,
  makeOutlinedButtonColorStyles,
  makeSizeStyles,
  makeTextButtonColorStyles,
} from './buttonStyles';

export { ButtonBaseTestId as ButtonTestId } from './ButtonBase';

export type ButtonProps = CommonButtonProps & {
  variant?: 'text' | 'outlined' | 'contained';
};

export default function Button(props: ButtonProps) {
  const {
    variant = 'text',
    type = 'secondary',
    size = 'large',
    ...restProps
  } = props;

  const isDarkTheme = false;
  const sizeStyles = useMemo(() => makeSizeStyles(size), [size]);
  const colorStyles = useMemo(() => {
    switch (variant) {
      case 'contained':
        return makeContainedButtonColorStyles(type, isDarkTheme);
      case 'outlined':
        return makeOutlinedButtonColorStyles(type, isDarkTheme);
      case 'text':
      default:
        return makeTextButtonColorStyles(type, isDarkTheme);
    }
  }, [isDarkTheme, type, variant]);

  const buttonStyles = useMemo(
    () => makeButtonStyles(sizeStyles, colorStyles),
    [sizeStyles, colorStyles],
  );

  return (
    <ButtonBase
      type={type}
      size={size}
      useTouchableOpacity={variant === 'text'}
      underlayColor={colorStyles.container.focused.backgroundColor}
      buttonStyles={buttonStyles}
      {...restProps}
    />
  );
}
