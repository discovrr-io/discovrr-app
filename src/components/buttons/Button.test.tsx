import React from 'react';

import { fireEvent, render } from '@testing-library/react-native';
import '@testing-library/jest-native';

import Button, { ButtonTestId } from './Button';

describe('Button', () => {
  const buttonTitle = 'My Button';

  describe('test large buttons', () => {
    describe('test large idle buttons', () => {
      it('renders an enabled large idle button', () => {
        const onPress = jest.fn();
        const { getByTestId, queryByTestId, toJSON } = render(
          <Button title={buttonTitle} onPress={onPress} />,
        );

        expect(queryByTestId(ButtonTestId.TEXT)).not.toBeNull();
        expect(queryByTestId(ButtonTestId.ACTIVITY_INDICATOR)).toBeNull();

        // expect(queryByTestId(ButtonTestId.TEXT)).toBeInTheDocument();
        // expect(queryByTestId(ButtonTestId.ACTIVITY_INDICATOR)).not.toBeInTheDocument();

        const buttonContainer = getByTestId(ButtonTestId.CONTAINER);
        fireEvent.press(buttonContainer);
        expect(onPress).toHaveBeenCalledTimes(1);

        // FIXME: This doesn't work
        // const buttonText = getByTestId(ButtonTestId.TEXT);
        // expect(buttonText.props.children).toContain(buttonTitle);

        expect(toJSON()).toMatchSnapshot();
      });
    });
  });
});
