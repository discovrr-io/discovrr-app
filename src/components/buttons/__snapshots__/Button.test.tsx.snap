// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test large buttons test large idle buttons renders an enabled large idle button 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "transparent",
      "borderColor": "transparent",
      "borderRadius": 12,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 16,
    }
  }
  testID="btn-container"
>
  <Text
    style={
      Array [
        Object {
          "color": "#141414",
          "fontFamily": "Rubik-Medium",
          "fontSize": 20,
        },
        undefined,
      ]
    }
    testID="btn-text"
  >
    My Button
  </Text>
</View>
`;
